crates/unleash-edge-metrics/src/send_unleash_metrics.rs

```diff
@@ -62,6 +64,125 @@ fn get_valid_token(token_cache: Arc<TokenCache>) -> Option<EdgeToken> {
+                                }
+                                StatusCode::FORBIDDEN | StatusCode::UNAUTHORIZED => {
+                                    MetricsSendError::Unauthed(format!(
+                                        "Upstream said we were not allowed to post metrics. Dropping this bucket to avoid consuming too much memory"
+                                    ))
+                                }
+                                StatusCode::TOO_MANY_REQUESTS => {
+                                    reinsert_batch(&metrics_cache, batch);
+                                    MetricsSendError::Backoff(format!(
+                                        "Upstream said it was too busy"
+                                    ))
+                                }
+                                StatusCode::INTERNAL_SERVER_ERROR
+                                | StatusCode::BAD_GATEWAY
+                                | StatusCode::SERVICE_UNAVAILABLE
+                                | StatusCode::GATEWAY_TIMEOUT => {
+                                    reinsert_batch(&metrics_cache, batch);
+                                    MetricsSendError::Backoff(format!(
+                                        "Upstream said it is struggling. It returned Http status {}",
+                                        status_code
+                                    ))
+                                }
+                                _ => {
+                                    reinsert_batch(&metrics_cache, batch);
+                                    MetricsSendError::Unknown(format!(
+                                        "Upstream returned an unexpected status code: {}",
+                                        status_code
+                                    ))
+                                }
+                            }
+                        }
+                        _ => {
+                            MetricsSendError::Unknown(format!(
+                                "Failed to send metrics: {edge_error:?}. Dropping this bucket to avoid consuming too much memory"
+                            ))
+                        }
+                    });
+                results.push(result);
+            }
+        }
+    }
+    results
+}
+
+pub fn create_once_off_send_metrics(
+    metrics_cache: Arc<MetricsCache>,
+    unleash_client: Arc<UnleashClient>,
+    token_cache: Arc<TokenCache>,
+) -> Pin<Box<dyn Future<Output = ()> + Send>> {
+    let metrics_cache = metrics_cache.clone();
+    let unleash_client = unleash_client.clone();
+    let token_cache = token_cache.clone();
+
+    Box::pin(async move {
+        let token = get_valid_token(token_cache.clone());
+        let Some(token) = token else {
+            warn!(
+                "No valid token found for final metrics send. Shutting down without flushing metrics."
+            );
+            return;
+        };
+        let envs = get_metrics_by_environment(&metrics_cache);
```

Comment:
Any of the metrics sending tasks can fail if they've been batched. We don't really care, we just want to display that list of fails to the user if there are any